Make a function that can take any non-negative integer as an argument and return it with its
digits in descending order. Essentially, rearrange the digits to create the highest possible
number. If no number is passed in, make the default value inputted = 111111.
Input = 43905 Output = 95430
Input = 584366 Output = 866543

In [1]:
def sort_number(n):
    if n == "":
        return 111111
    elif n > 0:
        n = str(n)
        digitlist = []
        for element in n:
            digitlist.append(element)
        digitlist.sort(reverse=True)
        joined = ''.join(digitlist)
        return joined
print(sort_number(43905))
print(sort_number(""))

95430
111111


Import jeopardy.csv and create a dataframe from the data. Find out the following:


In [31]:
import pandas as pd 
import numpy as np
df = pd.read_csv("JEOPARDY_CSV.csv",  delimiter=',')
df

Show Number    Air Date             Round  \
0              4680  12/31/2004         Jeopardy!   
1              4680  12/31/2004         Jeopardy!   
2              4680  12/31/2004         Jeopardy!   
3              4680  12/31/2004         Jeopardy!   
4              4680  12/31/2004         Jeopardy!   
...             ...         ...               ...   
216925         4999   5/11/2006  Double Jeopardy!   
216926         4999   5/11/2006  Double Jeopardy!   
216927         4999   5/11/2006  Double Jeopardy!   
216928         4999   5/11/2006  Double Jeopardy!   
216929         4999   5/11/2006   Final Jeopardy!   

                               Category    Value  \
0                               HISTORY    $200    
1       ESPN's TOP 10 ALL-TIME ATHLETES    $200    
2           EVERYBODY TALKS ABOUT IT...    $200    
3                      THE COMPANY LINE    $200    
4                   EPITAPHS & TRIBUTES    $200    
...                                 ...      ...   
216925                   RIDDLE ME THIS  $2,000    
216926                        "T" BIRDS  $2,000    
216927           AUTHORS IN THEIR YOUTH  $2,000    
216928                       QUOTATIONS  $2,000    
216929                   HISTORIC NAMES     None   

                                                 Question  \
0       For the last 8 years of his life, Galileo was ...   
1       No. 2: 1912 Olympian; football star at Carlisl...   
2       The city of Yuma in this state has a record av...   
3       In 1963, live on "The Art Linkletter Show", th...   
4       Signer of the Dec. of Indep., framer of the Co...   
...                                                   ...   
216925  This Puccini opera turns on the solution to 3 ...   
216926  In North America this term is properly applied...   
216927  In Penny Lane, where this "Hellraiser" grew up...   
216928  From Ft. Sill, Okla. he made the plea, Arizona...   
216929  A silent movie title includes the last name of...   

                                Answer  
0                           Copernicus  
1                           Jim Thorpe  
2                              Arizona  
3                           McDonald's  
4                           John Adams  
...                                ...  
216925                        Turandot  
216926                      a titmouse  
216927                    Clive Barker  
216928                        Geronimo  
216929  Grigori Alexandrovich Potemkin  

[216930 rows x 7 columns]

a. How many rows are there in the dataframe? What does a row represent?

There are 216930 rows and a row represents a jeopardy question , answer , show number,category, 
value and other relevant information.


In [32]:
df.describe()

Show Number
count  216930.000000
mean     4264.238519
std      1386.296335
min         1.000000
25%      3349.000000
50%      4490.000000
75%      5393.000000
max      6300.000000

In [33]:
df.head()


Show Number    Air Date      Round                         Category  Value  \
0         4680  12/31/2004  Jeopardy!                          HISTORY  $200    
1         4680  12/31/2004  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200    
2         4680  12/31/2004  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200    
3         4680  12/31/2004  Jeopardy!                 THE COMPANY LINE  $200    
4         4680  12/31/2004  Jeopardy!              EPITAPHS & TRIBUTES  $200    

                                            Question      Answer  
0  For the last 8 years of his life, Galileo was ...  Copernicus  
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe  
2  The city of Yuma in this state has a record av...     Arizona  
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's  
4  Signer of the Dec. of Indep., framer of the Co...  John Adams

In [34]:
df.columns

Index(['Show Number', ' Air Date', ' Round', ' Category', ' Value',
       ' Question', ' Answer'],
      dtype='object')

In [35]:
df.dtypes


Show Number     int64
 Air Date      object
 Round         object
 Category      object
 Value         object
 Question      object
 Answer        object
dtype: object

In [37]:
import datetime
import pandas as pd
import numpy as np
import random
df["Air Date"] = pd.to_datetime(df["Air Date"])

KeyError: 'Air Date'

In [28]:
tday = datetime.date.today()
print(tday.year)    

2021


In [42]:
df.columns

Index(['Show Number', ' Air Date', ' Round', ' Category', ' Value',
       ' Question', ' Answer'],
      dtype='object')

In [46]:
df.columns = df.columns.str.replace(' ', '')
print(df.columns)

Index(['ShowNumber', 'AirDate', 'Round', 'Category', 'Value', 'Question',
       'Answer'],
      dtype='object')


In [51]:
import datetime
import pandas as pd
import numpy as np
import random
df["AirDate"] = pd.to_datetime(df["AirDate"])
df["AirDate"].head(4)

0   2004-12-31
1   2004-12-31
2   2004-12-31
3   2004-12-31
Name: AirDate, dtype: datetime64[ns]

In [52]:
df.dtypes

ShowNumber             int64
AirDate       datetime64[ns]
Round                 object
Category              object
Value                 object
Question              object
Answer                object
dtype: object

In [53]:
df.AirDate.dt.year

0         2004
1         2004
2         2004
3         2004
4         2004
          ... 
216925    2006
216926    2006
216927    2006
216928    2006
216929    2006
Name: AirDate, Length: 216930, dtype: int64

In [65]:
group_names = df.groupby("AirDate")
#df.AirDate.dt.year
group_names["ShowNumber"].unique()

AirDate
1984-09-10       [1]
1984-09-11       [2]
1984-09-12       [3]
1984-09-13       [4]
1984-09-14       [5]
               ...  
2012-01-20    [6295]
2012-01-23    [6296]
2012-01-24    [6297]
2012-01-25    [6298]
2012-01-27    [6300]
Name: ShowNumber, Length: 3640, dtype: object

In [69]:
#create a column for the year aired
df['Year'] = pd.DatetimeIndex(df['AirDate']).year
df.head()

ShowNumber    AirDate      Round                         Category  Value  \
0        4680 2004-12-31  Jeopardy!                          HISTORY  $200    
1        4680 2004-12-31  Jeopardy!  ESPN's TOP 10 ALL-TIME ATHLETES  $200    
2        4680 2004-12-31  Jeopardy!      EVERYBODY TALKS ABOUT IT...  $200    
3        4680 2004-12-31  Jeopardy!                 THE COMPANY LINE  $200    
4        4680 2004-12-31  Jeopardy!              EPITAPHS & TRIBUTES  $200    

                                            Question      Answer  Year  
0  For the last 8 years of his life, Galileo was ...  Copernicus  2004  
1  No. 2: 1912 Olympian; football star at Carlisl...  Jim Thorpe  2004  
2  The city of Yuma in this state has a record av...     Arizona  2004  
3  In 1963, live on "The Art Linkletter Show", th...  McDonald's  2004  
4  Signer of the Dec. of Indep., framer of the Co...  John Adams  2004

In [77]:
df['Year'] = pd.DatetimeIndex(df['AirDate']).year
df.head()
group_names = df.groupby("Year")
group_names["ShowNumber"].nunique()


Year
1984     23
1985     16
1986     25
1987     22
1988     22
1989     36
1990     73
1991     24
1992     32
1993     35
1994     19
1995     19
1996     83
1997    221
1998    223
1999    229
2000    225
2001    203
2002    114
2003    158
2004    220
2005    225
2006    230
2007    231
2008    232
2009    226
2010    231
2011    224
2012     19
Name: ShowNumber, dtype: int64

In [83]:
df["Category"].value_counts()

BEFORE & AFTER         547
SCIENCE                519
LITERATURE             496
AMERICAN HISTORY       418
POTPOURRI              401
                      ... 
TOURISM                  1
BURIED AT ARLINGTON      1
HISTORIC MESSAGES        1
FINANCE HISTORY          1
THE BOOK TRADE           1
Name: Category, Length: 27983, dtype: int64

In [90]:
df["Month"] = pd.DatetimeIndex(df['AirDate']).month
df


ShowNumber    AirDate             Round  \
0             4680 2004-12-31         Jeopardy!   
1             4680 2004-12-31         Jeopardy!   
2             4680 2004-12-31         Jeopardy!   
3             4680 2004-12-31         Jeopardy!   
4             4680 2004-12-31         Jeopardy!   
...            ...        ...               ...   
216925        4999 2006-05-11  Double Jeopardy!   
216926        4999 2006-05-11  Double Jeopardy!   
216927        4999 2006-05-11  Double Jeopardy!   
216928        4999 2006-05-11  Double Jeopardy!   
216929        4999 2006-05-11   Final Jeopardy!   

                               Category    Value  \
0                               HISTORY    $200    
1       ESPN's TOP 10 ALL-TIME ATHLETES    $200    
2           EVERYBODY TALKS ABOUT IT...    $200    
3                      THE COMPANY LINE    $200    
4                   EPITAPHS & TRIBUTES    $200    
...                                 ...      ...   
216925                   RIDDLE ME THIS  $2,000    
216926                        "T" BIRDS  $2,000    
216927           AUTHORS IN THEIR YOUTH  $2,000    
216928                       QUOTATIONS  $2,000    
216929                   HISTORIC NAMES     None   

                                                 Question  \
0       For the last 8 years of his life, Galileo was ...   
1       No. 2: 1912 Olympian; football star at Carlisl...   
2       The city of Yuma in this state has a record av...   
3       In 1963, live on "The Art Linkletter Show", th...   
4       Signer of the Dec. of Indep., framer of the Co...   
...                                                   ...   
216925  This Puccini opera turns on the solution to 3 ...   
216926  In North America this term is properly applied...   
216927  In Penny Lane, where this "Hellraiser" grew up...   
216928  From Ft. Sill, Okla. he made the plea, Arizona...   
216929  A silent movie title includes the last name of...   

                                Answer  Year  Month  
0                           Copernicus  2004     12  
1                           Jim Thorpe  2004     12  
2                              Arizona  2004     12  
3                           McDonald's  2004     12  
4                           John Adams  2004     12  
...                                ...   ...    ...  
216925                        Turandot  2006      5  
216926                      a titmouse  2006      5  
216927                    Clive Barker  2006      5  
216928                        Geronimo  2006      5  
216929  Grigori Alexandrovich Potemkin  2006      5  

[216930 rows x 9 columns]

In [96]:
group_names = df.groupby("Month")
group_names["ShowNumber"].max()

Month
1     6300
2     6096
3     6119
4     6140
5     6162
6     6184
7     6205
9     6215
10    6236
11    6258
12    6280
Name: ShowNumber, dtype: int64